generators/argonaut-r4/generator.rb

```diff
@@ -0,0 +1,504 @@
+  end
+end
+
+def generate_sequence(sequence)
+  puts "Generating #{sequence[:name]}\n"
+  file_name = OUT_PATH + '/us_core_r4/'+ sequence[:name].downcase + '_sequence.rb'
+
+  template = ERB.new(File.read('./templates/sequence.rb.erb'))
+  output =   template.result_with_hash(sequence)
+  unless File.directory?(OUT_PATH + '/us_core_r4')
+    FileUtils.mkdir_p(OUT_PATH + '/us_core_r4')
+  end
+  File.write(file_name, output)
+end
+
+def get_search_param_json(resource, param)
+  begin
+    param = "id" if param == "_id" # special case for _id
+    uri = "https://build.fhir.org/ig/HL7/US-Core-R4/SearchParameter-us-core-#{resource}-#{param}.json"
+    get_json_from_uri(uri)
+  rescue
+  end
+end
+
+def get_variable_type_from_structure_def(resource, profile, var)
+  resource_struct_def = get_json_from_uri(profile)
+  element_def = resource_struct_def['snapshot']['element'].select{|el| el['id'] == "#{resource}.#{var}"}.first if !resource_struct_def.nil?
+  return element_def['type'].first['code'] if !element_def.nil?
+  return 'test'
+end
+
+def get_variable_contains_multiple(resource, profile, var)
+  resource_struct_def = get_json_from_uri(profile)
+  element_def = resource_struct_def['snapshot']['element'].select{|el| el['id'] == "#{resource}.#{var}"}.first if !resource_struct_def.nil?
+  return element_def['max'] == "*" if !element_def.nil?
+end
+
+def create_authorization_test(sequence)
+  authorization_test = {
+    tests_that: "Server rejects #{sequence[:resource]} search without authorization",
+    index:  '%02d' % (sequence[:tests].length + 1),
+    link: "http://www.fhir.org/guides/argonaut/r2/Conformance-server.html"
+  }
+
+  authorization_test[:test_code] = %(
+        @client.set_no_auth
+        skip 'Could not verify this functionality when bearer token is not set' if @instance.token.blank?
+
+        reply = get_resource_by_params(versioned_resource_class('#{sequence[:resource]}'), {patient: @instance.patient_id})
+        @client.set_bearer_token(@instance.token)
+        assert_response_unauthorized reply
+  )
+
+  sequence[:tests] << authorization_test
+end
+
+def create_search_test(sequence, search_param)
+  search_test = {
+    tests_that: "Server returns expected results from #{sequence[:resource]} search by #{search_param[:names].join('+')}",
+    index: '%02d' % (sequence[:tests].length + 1),
+    link: "https://build.fhir.org/ig/HL7/US-Core-R4/CapabilityStatement-us-core-server.html"
+  }
+
+  is_first_search = search_test[:index] == '02' # if first search - fix this check later
+  if is_first_search then
+    search_test[:test_code] = %(
+        #{get_search_params(sequence[:resource], sequence[:profile], search_param[:names])}
+        reply = get_resource_by_params(versioned_resource_class('#{sequence[:resource]}'), search_params)
+        assert_response_ok(reply)
+        assert_bundle_response(reply)
+
+        resource_count = reply.try(:resource).try(:entry).try(:length) || 0
```

Comment:
Do we want to use `#try` or the `&.` here?

Reply:
We should get rid of the try's, but let's split this update out into the next PR.